lib/arjdbc/abstract/statement_cache.rb

```diff
@@ -0,0 +1,58 @@
+require 'active_record/connection_adapters/statement_pool'
+
+module ArJdbc
+  module Abstract
+    module StatementCache
+
+      # This works a little differently than the AR implementation in that
+      # we are storing an actual PreparedStatement object instead of just
+      # the name of the prepared statement
+      class StatementPool < ActiveRecord::ConnectionAdapters::StatementPool
+
+        private
+
+        def dealloc(statement)
+          statement.close
+        end
+
+      end
+
+      def initialize(connection, logger, config)
+        super
+
+        # Only say we support the statement cache if we are using prepared statements
+        # and have a max number of statements defined
+        statement_limit = self.class.type_cast_config_to_integer(config[:statement_limit])
+        @jdbc_statement_cache_enabled = config[:prepared_statements] && (statement_limit.nil? || statement_limit > 0)
+
+        @statements = StatementPool.new(statement_limit) # AR (5.0) expects this to be stored as @statements
+      end
+
+      # Clears the prepared statements cache.
+      def clear_cache!
+        @statements.clear
+      end
+
+      def delete_cached_statement(sql)
+        @statements.delete(cached_statement_key(sql))
+      end
+
+      def fetch_cached_statement(sql)
+        @statements[cached_statement_key(sql)] ||= @connection.connection.prepare_statement(sql)
```

Comment:
I had originally created a method in RubyJdbcConnection.java to handle the statement preparation, but realized it wasn't really adding anything and meant more code to maintain. I can go back to using that if you think we should avoid directly referencing the connection object here.

Here's the method I had originally added:

```
    @JRubyMethod(name = "prepare_statement", required = 1)
    public IRubyObject prepareStatement(final ThreadContext context, final IRubyObject sql) throws SQLException {
        return withConnection(context, new Callable<IRubyObject>() {
            public IRubyObject call(final Connection connection) throws SQLException {
                final String query = sql.convertToString().getUnicodeValue();
                final PreparedStatement statement = connection.prepareStatement(query);
                return JavaEmbedUtils.javaToRuby(context.getRuntime(), statement);
            }
        });
    }
```